crypto/src/vrf/schnorrkel/mod.rs

```diff
@@ -0,0 +1,237 @@
+// Copyright (c) 2021 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://spdx.org/licenses/MIT
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Author(s): S. Afach
+
+use crate::random::{CryptoRng, Rng};
+use merlin::Transcript;
+use schnorrkel::Keypair;
+use serialization::{Decode, Encode};
+
+use self::data::SchnorrkelVRFReturn;
+
+use super::{primitives::VRFReturn, VRFError};
+
+const PUBKEY_LEN: usize = 32;
+const PRIVKEY_LEN: usize = 64; // scalar + nonce
+
+pub mod data;
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[must_use]
+pub struct SchnorrkelPublicKey {
+    key: schnorrkel::PublicKey,
+}
+
+impl SchnorrkelPublicKey {
+    pub fn from_private_key(private_key: &SchnorrkelPrivateKey) -> Self {
+        SchnorrkelPublicKey {
+            key: private_key.key.to_public(),
+        }
+    }
+
+    pub fn verify_generic_vrf_data(
+        &self,
+        message: Transcript,
+        vrf_data: &VRFReturn,
+    ) -> Result<(), VRFError> {
+        match vrf_data {
+            VRFReturn::Schnorrkel(d) => self.verify_vrf_data(message, d),
+        }
```

Comment:
What is this supposed to do if another enum variant is added to `VRFReturn`?

Reply:
Just to be clear, are you proposing doing something specific? So far everything is covered because of match.

Reply:
I know you don't like matching on tuples and wildcards but I'd find something like the following pattern quite clear:
```rust
fn verify_vrf(vrf_pk: PublicKey, msg: Transcript, vrf_return: VRFReturn) -> Result<(), Error> {
  match (vrf_pk, vrf_return) {
    (PublicKey::Schnorrkel(pk), VRFReturn::Schnorrkel(ret)) => pk.verify_vrf_data(msg, ret),
    (PublicKey::SomeOther(pk), VRFReturn::SomeOther(ret)) => pk.verify_vrf_data(msg, ret),
    (pk, ret) => {
      assert_ne!(pk.kind(), ret.kind(), "VRF kind {:?} not covered", pk.kind());
      Err(Error::VRFKindMismatch(pk.kind(), ret.kind()))
    }
  }
}
```
(`vrf_pk` could be `self` to make it a method on PublicKey)

Reply:
> I know you don't like matching on tuples and wildcards

How well you know me 😄 

I get the appeal, but I would rather have the compiler yell at me if I add a new type and forget to cover it instead of having this deceiving beauty that can lead to bugs in the future.